### DISCUSSION TITLE
Duplicating all of an entity's components

### DISCUSSION THREAD

**Discussion by WarEagle451** [skypjack/entt#684] [Q&A]
The straight forward way of doing this would be to check if the entity has the component, then if so get it and copy all it's members into a new component. The issue with this way of duplicating an entity's components is that it could require a mountain of code to be written depending on how many different components could be used. Is there a better alternative for duplicating an entity's components? If not what is the best way to manage such an issue?

**Comment by skypjack** [comment]
`EnTT` makes heavy use of type erasure to avoid forcing users to announce their types. On the other hand, even the registry does not have a clue of the types of components used with it over time.
Because of this, technically speaking there is no way to work with a component without specifying its type, unless you _prepare_ the right amount of erased functions to accompany an erased pool. By default, the library doesn't offer these functions. This is because the desired behavior often depends on the application and you would find yourself paying for all the instantiations required to support all the feature you don't need.

So, am I saying there is no solution for this? Quite the opposite. 🙂 
But you have to dig a bit into the library and define what you need, what you want.

`EnTT` offers what is called a [poly storage](https://github.com/skypjack/entt/wiki/Crash-Course:-entity-component-system#meet-the-runtime), that is, literally an empty class that you can design as you like for all types or on a per-type basis to add _opaque functionalities_ to your pools with the desired granularity.
Something I've to document properly and to refine, but still. [Here](https://godbolt.org/z/h4q9o1ecj) it is an example that copies the components of an entities to another entity. [This](https://github.com/skypjack/entt/blob/master/test/entt/entity/poly_storage.cpp) is a slightly more complex example with more functionalities instead.
It's also quite easy to define a poly storage that supports copying entities and components between registries that might also have different types for their entity identifiers. Since you can (and must actually) customize everything, it goes without saying that you can do whatever you want. 🤷‍♂️ 

Most likely, I'll provide some pre-cooked implementations in future. Though, I'm currently working on other things because of my job, so consider this a low prio task at the moment.

**Reply by WarEagle451** [reply]
Thank you so much for the help, I had been stuck on this for quite some time now

**Reply by skypjack** [reply]
You're welcome @WarEagle451 have you been able to duplicate entities opaquely?

**Reply by WarEagle451** [reply]
I haven't got around to implementing it yet, it's a hobby project so it might be a while before I get around to it. Hopefully I'll remember to update you when I do it

**Comment by keithlostracco** [comment]
Is there an updated version of the example you linked above? 

It appears that as of EnTT3.9 poly storage has been fully removed...

**Reply by skypjack** [reply]
```cpp
for(auto [id, storage]: registry.storage()) {
    if(storage.contains(src)) {
        storage.push(dst, storage.value(src));
    }
}
```
Where `src` and `dst` are entities and all component types are at least copyable.

**Reply by keithlostracco** [reply]
Sweet! That's pretty straightforward now. Thanks.

**Reply by keithlostracco** [reply]
It turns out I was already using `storage.push()`  (with just one arg for the destination entity) to create components so it would have the same components when I later wrote those values from a serialize function (for Undo/Redo). I didn't realize when reading the documentation for `storage.push()` that:

> elem Optional opaque element to forward to mixins, if any.

meant I could pass the opaque values of another entity's matching component to set the values of the destination entity. Apologies, I should have read that more carefully.

**Reply by skypjack** [reply]
No worries 🙂 I'm glad it worked. 👍

**Reply by khalv** [reply]
Neither storage.push nor storage.value are defiend for me... what branch of entt does this apply to?

**Reply by skypjack** [reply]
`master` since a while ago. What version of EnTT are you using @khalv ?

**Reply by mergipe** [reply]
> ```c++
> for(auto [id, storage]: registry.storage()) {
>     if(storage.contains(src)) {
>         storage.push(dst, storage.value(src));
>     }
> }
> ```
> 
> Where `src` and `dst` are entities and all component types are at least copyable.

Hi. Is there any way of doing this copy between two different registries? I have a registry populated with template entities (prefabs) and I'm looking for a way to copy all of a template's components to a new entity in the "world" registry in a clean way. The problem with this approach is that the world registry maybe won't have all the storages needed to make the copy.

**Reply by Avelina9X** [reply]
@mergipe yes this should work between two registries no problem, however this requires that both the source registry and destination registry have the storages already created for all components. You would iterate over all registry storages in the prefab registry getting the storage pointer and component `id` hash, then you'd use `id` to obtain the type erased storage from the world registry, and push the value into the storage. If the world registry does not already have a storage for that `id` already created, it would return a nullptr.

If you look at the cross-registry copy example, you can see a trick that was used via RTTI to automatically create unallocated storages for you; https://github.com/skypjack/entt/blob/1333fa53129e7cfded5a9640c4336a254049917b/test/example/entity_copy.cpp#L39-L49

Personally I'm less of a fan of this trick, but this is pretty much the only way of doing it without defining ahead of time which specific components needs to be copied using some compile time type list.

**Reply by skypjack** [reply]
You can also use a storage mixin to achieve the same. The idea is that you get the storage from the source registry and invoke the `bind` function with the _new_ registry. The mixin can _initialize_ the new registry (i.e. with call to `storage<Type>()`), then forward the request to the newly created pool instead of its own base class.

**Reply by skypjack** [reply]
CC @mergipe @Avelina9X (in case you are looking for a different solution to the same problem in your code)
